File: publish-results-pr/internal/processing/processor.go
package processing

import (
	"fmt"
	"log"

	"ahk/publishresultpr/internal/appargs"
	"ahk/publishresultpr/internal/processing/imagefilesfinder"
	"ahk/publishresultpr/internal/processing/neptunparser"
	resultsfileparser "ahk/publishresultpr/internal/processing/resultsfile"
)

type Processor interface {
	Process(appArgs appargs.AppArgs, workingDir string) (res AhkProcessResult, err error)
}

type processor struct {
}

func NewProcessor() Processor {
	return new(processor)
}

func (p *processor) Process(appArgs appargs.AppArgs, workingDir string) (res AhkProcessResult, err error) {
	neptunParser := new(neptunparser.NeptunParser)
	neptun, err := neptunParser.ParseNeptun(appArgs.NeptunFileName)
	if err != nil {
		return AhkProcessResult{}, err
	}
	log.Printf("Neptun: %s\n", *neptun)

	imageFilesFinder := new(imagefilesfinder.ImageFileFinder)
	imageFiles, err := imageFilesFinder.GetImageFiles(workingDir, appArgs.ImageExtension)
	if err != nil {
		return AhkProcessResult{}, err
	}
	for _, imageFile := range imageFiles {
		log.Printf("Found image file: %s\n", imageFile)
	}

	resultsFileParser := new(resultsfileparser.ResultsFileParser)
	results, err := resultsFileParser.ParseResultsFile(appArgs.ResultFile)
	if err != nil {
		return AhkProcessResult{}, err
	}
	for _, r := range results {
		log.Printf("Result for %s %s %f\n", r.ExerciseName, r.TaskName, r.Points)
	}

	return AhkProcessResult{
		GitHubRepoName:       appArgs.GitHubRepoFullName,
		GitHubBranch:         appArgs.GitHubBranch,
		GitHubPullRequestNum: appArgs.GitHubPullRequestNum,
		GitHubCommitHash:     appArgs.GitCommitHash,
		NeptunCode:           *neptun,
		ImageFiles:           imageFiles,
		Result:               results,
		Origin:               fmt.Sprintf("https://github.com/%s/commit/%s https://github.com/%s/actions/runs/%s", appArgs.GitHubRepoFullName, appArgs.GitCommitHash, appArgs.GitHubRepoFullName, appArgs.GitHubActionRunId),
	}, nil
}

File: publish-results-pr/Processing/Processor.cs
using System.Text.Json;

using PublishResult.AppArgs;

namespace PublishResult.Processing;

public readonly struct AhkProcessResult
{
    public string GitHubRepoName { get; init; }
    public string GitHubBranch { get; init; }
    public int GitHubPullRequestNum { get; init; }
    public string GitHubCommitHash { get; init; }
    public string NeptunCode { get; init; }
    public string?[]? ImageFiles { get; init; }
    public string Result { get; init; }
    public string Origin { get; init; }
}

public static class Processor
{
    public static AhkProcessResult Process(AppArgs.AppArgs appArgs, string workDir)
    {
        var neptun = NeptunParser.ParseNeptun(appArgs.NeptunFileName);

        Console.WriteLine($"Neptun: {neptun}");

        string?[]? imageFiles = null;
        if (appArgs.ImageExtension == "")
            Console.WriteLine("Image files gathering not requested");
        else
        {
            imageFiles = ImageFilesFinder.GetImageFiles(workDir, appArgs.ImageExtension);
            if (imageFiles != null && imageFiles.Length > 0)
            {
                foreach (var imageFile in imageFiles)
                {
                    Console.WriteLine($"Found image file: {imageFile}");
                }
            }
            else
                Console.WriteLine("Found 0 image files");
        }

        string result = null!;
        if (appArgs.ResultFile == "")
            Console.WriteLine("Result file processing not requested");
        else
        {
            // JSON
            result = File.ReadAllText(appArgs.ResultFile);
            var exercises = JsonSerializer.Deserialize<Exercise[]>(result);
            if (exercises != null)
                foreach (var exercise in exercises)
                {
                    Console.WriteLine(exercise);
                }
        }

        return new AhkProcessResult
        {
            GitHubRepoName = appArgs.GitHubRepoFullName,
            GitHubBranch = appArgs.GitHubBranch,
            GitHubPullRequestNum = appArgs.GitHubPullRequestNum,
            GitHubCommitHash = appArgs.GitCommitHash,
            NeptunCode = neptun!,
            ImageFiles = imageFiles,
            Result = result,
            Origin = $"https://github.com/{appArgs.GitHubRepoFullName}/commit/{appArgs.GitCommitHash} https://github.com/{appArgs.GitHubRepoFullName}/actions/runs/{appArgs.GitHubActionRunId}",
        };
    }
}
